### PR TITLE
feat: populate standard audit log field for justification(java client)

### PR DESCRIPTION
**Pull request by sqin2019**
java client side change for issue #266

### CODE REVIEW

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java [hunk 1]
```diff
@@ -121,6 +126,20 @@ public AuditLog.Builder auditLogBuilderWithJustification(
           Value.newBuilder().setStructValue(justificationStruct).build());
       auditLogBuilderCopy.setMetadata(metadataBuilder.build());
 
+      Claim justificationsClaim = jwt.getClaim("justs");
+      if (justificationsClaim.isNull()) {
+        log.warn("can't find 'justs' in claims");
+      } else {
+        RequestMetadata.Builder requestMetadataBuilder =
+            auditLogBuilderCopy.getRequestMetadataBuilder();
```
Comment (capri-xiyue) [inline clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:134]: I saw that line 123 uses `auditLogBuilderCopy.getMetadata().toBuilder()`. Does requestMetadata has the same function `auditLogBuilderCopy.getRequestMetadat().toBuilder()`? If there is, what's the difference?

File: clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java [hunk 2]
```diff
@@ -78,14 +92,31 @@ public void processShouldPopulateJustificationMetadata() throws Exception {
...
     Struct wantMetadata =
         Struct.newBuilder()
             .putFields(
                 "justification", Value.newBuilder().setStructValue(wantJustification).build())
             .build();
+    RequestMetadata wantRequestMetadata =
+        RequestMetadata.newBuilder()
+            .setRequestAttributes(
+                Request.newBuilder()
+                    .setReason("[{\"category\":\"explanation\",\"value\":\"need-access\"}]")
```
Comment (capri-xiyue) [inline clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:112]: Does `[{"category":"explanation","value":"need-access"}]` work?

Comment (sqin2019) [reply clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:112]: No, I tried that, not sure why.

Comment (capri-xiyue) [reply clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:112]: What's the error message?

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java (hunk 1 above)
Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java]: GetXXBuilder() or getXX().toBuilder() will return a default/empty instance if it is null. The different between the two method is one returns a Builder one returns an object, use getXXBuilder() since it is all we need otherwise will need to convert the object to builder to set values in it.

The logic is what you mentioned, it only adds the reason and does not alter anything else when justs is found and non0null/empty.

File: clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java (hunk 2 above)
Comment (sqin2019) [reply clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:112]:
```
java.lang.Error: Unresolved compilation problems: 
 The method setReason(String) in the type AttributeContext.Request.Builder is not applicable for the arguments (String, String)
 Syntax error on tokens, delete these tokens
```
looks like it still need escape for "" inside the sentence.

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java [hunk 3]
```diff
@@ -121,6 +127,26 @@ public AuditLog.Builder auditLogBuilderWithJustification(
           Value.newBuilder().setStructValue(justificationStruct).build());
       auditLogBuilderCopy.setMetadata(metadataBuilder.build());
 
+      Claim justificationsClaim = jwt.getClaim("justs");
```
Comment (capri-xiyue) [inline clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: Can we add a comment that the logic here is no justs and then continue with other the populate of other fields?
Context: https://github.com/abcxyz/lumberjack/pull/293#discussion_r971292134

Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: I have added a change to return early when justs is not found or null or empty, do you think it is enough?

Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: Ok, I just see your comment there, for the concern that missing populating other new fields, I think unit test will catch that, what do you think?

Comment (capri-xiyue) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: Why the unit tests will catch it?For example, later we when want to populate another field called "C", the developer need to know the logic is "C" filed still need to get populated even though there is no justs found. The developer need to know the logic and then modify the UT to catch that. Our current UT won't be able to catch it.

Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: Right, current UT should only cover existing functionalities, if adding a new field, the developer should add UT to test that is what I meant.

Comment (capri-xiyue) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: For the developers to add UT, they need to know the business logic is to continue when there is no justs found. That's why I asked to add a comment. Currently the code looks unclear whether the logic is to exit or to continue.

Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: Added to avoid future confusions, please check newest commit, thanks!

Comment (capri-xiyue) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: Can we add a Java comments instead of adding it in the logging?

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java (hunk 1 above)
Comment (capri-xiyue) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java]: ack. Thank you for the explanation and confirmation!

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java (hunk 3 above)
Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: Ack.

**Review by capri-xiyue — APPROVED** [review]

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java [hunk 4]
```diff
@@ -121,6 +127,29 @@ public AuditLog.Builder auditLogBuilderWithJustification(
...
+
+      RequestMetadata.Builder requestMetadataBuilder =
+          auditLogBuilderCopy.getRequestMetadataBuilder();
+      Request.Builder requestAttributesBuilder =
+          requestMetadataBuilder
+              .getRequestAttributesBuilder()
+              .setReason(new Gson().toJson(justificationsList));
+      requestMetadataBuilder =
+          requestMetadataBuilder.setRequestAttributes(requestAttributesBuilder);
+      auditLogBuilderCopy.setRequestMetadata(requestMetadataBuilder);
```
Comment (yolocs) [inline clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:152]: What if there is an existing requestMetadata in the audit log?

**Review by yolocs — COMMENTED** [review]
cc: @myurtoglu for a quick java readability review

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java (hunk 4 above)
Comment (capri-xiyue) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:152]: Context: https://github.com/abcxyz/lumberjack/pull/304#discussion_r972475115

Comment (raserva) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:152]: I think its worth adding a unit test for this.

File: clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java [hunk 5]
```diff
@@ -78,14 +92,31 @@ public void processShouldPopulateJustificationMetadata() throws Exception {
...
+                    .build())
```
Comment (raserva) [inline clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:113]: We're missing negative tests - validate that the request metadata isn't populated on null/empty justs

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java [hunk 6]
```diff
@@ -121,6 +127,29 @@ public AuditLog.Builder auditLogBuilderWithJustification(
           Value.newBuilder().setStructValue(justificationStruct).build());
       auditLogBuilderCopy.setMetadata(metadataBuilder.build());
 
+      Claim justificationsClaim = jwt.getClaim("justs");
```
Comment (raserva) [inline clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: better to have a constant for this string

Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:130]: ack.

File: clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java (hunk 4 above)
Comment (sqin2019) [reply clients/java-logger/library/src/main/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessor.java:152]: I will go back to the common way of get the object instead of builder since multiple people are having doubts on it.

File: clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java (hunk 5 above)
Comment (sqin2019) [reply clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:113]: We have a ut setLogJustificationShouldPopulateJustificationMetadata in which request metadata is skipped when justs is null, I will create seperate ut when list is null or empty, thanks for pointing it out!

File: clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java [hunk 7]
```diff
@@ -78,14 +93,31 @@ public void processShouldPopulateJustificationMetadata() throws Exception {
...
     Struct wantMetadata =
         Struct.newBuilder()
             .putFields(
                 "justification", Value.newBuilder().setStructValue(wantJustification).build())
             .build();
+    RequestMetadata wantRequestMetadata =
+        RequestMetadata.newBuilder()
+            .setRequestAttributes(
+                Request.newBuilder()
+                    .setReason("[{\"category\":\"explanation\",\"value\":\"need-access\"}]")
```
Comment (yolocs) [inline clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:113]: @capri-xiyue Can you confirm on the Go side we're handling it as a list as well?

**Review by yolocs — APPROVED** [review]

File: clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java (hunk 7 above)
Comment (sqin2019) [reply clients/java-logger/library/src/test/java/com/abcxyz/lumberjack/auditlogclient/processor/JustificationProcessorTests.java:113]: Checked JVS java client, it is also using the [justs claim as a list](https://github.com/abcxyz/jvs/blob/e718d4664467b880991b8e2a400070c2aa93a0b9/client-lib/java/src/main/java/com/abcxyz/jvs/JvsClient.java#L76). Also tried to convert the justs Claim directly to json using both gson and jackson, neither of them worked because Claim holds the value in a generic way.

**Review by raserva — APPROVED** [review]
